### PR TITLE
Login

### PR DESCRIPTION
**Pull request by TheoPth**
Ajout du login et private routes

### CODE REVIEW

File: src/prototypes/Student.prototype.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    this.id = idStudent;
+    this.email = emailStudent;
+    this.role = roleStudent;
+    this.firstName = firstNameStudent;
+    this.lastName = lastNameStudent;
+    this.class = classStudent;
+  }
+
+  static dbToStudent(obj) {
+    const course = new Student(obj.idStudent, obj.emailStudent, obj.roleStudent, obj.firstNameStudent, obj.lastNameStudent, obj.classStudent);
```
Comment (yannick-mayeur) [inline src/prototypes/Student.prototype.js]: In the database the attributes are all lowercase. You should use idstudent, emailstudent, etc.

Comment (yannick-mayeur) [reply src/prototypes/Student.prototype.js]:
```
const course = new Student(obj.idstudent, obj.emailstudent, obj.rolestudent, obj.firstnamestudent, obj.lastnamestudent, obj['class-student']);
```

File: src/routes/connexion.route.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+      });
+    } else {
+      res.statusMessage = 'You are not allowed to create an account with this email.';
+      res.status(401).send();
+    }
+  });
+
+  router.get('/login/isTokenValid', async (req, res) => {
+    M.Login.isTokenValid(req.body.token).then((bool) => {
+      res.status(200).send({ isValid: bool });
```
Comment (yannick-mayeur) [inline src/routes/connexion.route.js]: I don't quite understand why but this always return `{}`.

File: src/prototypes/Teacher.prototype.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  constructor(idTeacher, emailTeacher, roleTeacher, firstNameTeacher, lastNameTeacher) {
+    this.id = idTeacher;
+    this.email = emailTeacher;
+    this.role = roleTeacher;
+    this.firstName = firstNameTeacher;
+    this.lastName = lastNameTeacher;
+  }
+
+  static dbToTeacher(obj) {
+    const course = new Teacher(obj.idTeacher, obj.emailTeacher, obj.roleTeacher, obj.firstNameTeacher, obj.lastNameTeacher);
```
Comment (yannick-mayeur) [inline src/prototypes/Teacher.prototype.js]: Same problem there was with the Student prototype.

File: src/routes/connexion.route.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        res.statusMessage = err;
+        res.status(500).send();
+      });
+    } else {
+      res.statusMessage = 'You are not allowed to create an account with this email.';
+      res.status(401).send();
+    }
+  });
+
+  router.get('/login/isTokenValid', async (req, res) => {
```
Comment (yannick-mayeur) [inline src/routes/connexion.route.js]: Is it possible for a GET request to have a body?

**Comment by yannick-mayeur** [comment]
Some little problems in the PR but overall it is a very good job, this will really help everyone continue their respective work.

File: src/routes/connexion.route.js (hunk 2 above)
Comment (yannick-mayeur) [reply src/routes/connexion.route.js]: This version of isTokenValid seems to solve the problem.
```
  async isTokenValid(token) {
    try {
      jwt.verify(token, 'secret');
      return true;
    } catch(err) {
      return false;
    }
  }
```

Comment (TheoPth) [reply src/routes/connexion.route.js]: I forget the return statement, fixed it.

File: src/routes/connexion.route.js (hunk 4 above)
Comment (TheoPth) [reply src/routes/connexion.route.js]: Not in our case

Comment (yannick-mayeur) [reply src/routes/connexion.route.js]: So should this be a post ?

File: src/models/login.model.js [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+  async isTokenValid(token) {
+    return jwt.verify(token, 'secret', (err) => {
+      return err == undefined;
+    });
+  },
+  generateToken(id) {
+    // TODO secret in session
+    // jwt.sign({ id, role }, process.env.SESSION_SECRET, {
+    //     expiresIn: 86400, // expires in 24 hours
+    // });
```
Comment (yannick-mayeur) [inline src/models/login.model.js]: If you change this I think we are ready to merge !

File: src/routes/connexion.route.js (hunk 4 above)
Comment (yannick-mayeur) [reply src/routes/connexion.route.js]: Just saw your commit. I'm resolving this.

File: src/models/login.model.js [hunk 6]
```diff
@@ -108,12 +108,12 @@ const login = {
...
   },
   async isTokenValid(token) {
-    return jwt.verify(token, 'secret', (err) => {
+    return jwt.verify(token, process.env.SESSION_SECRET="secret", (err) => {
       return err == undefined;
     });
   },
   generateToken(id) {
-  return jwt.sign({ id }, 'secret', {
+  return jwt.sign({ id }, process.env.SESSION_SECRET="secret", {
```
Comment (yannick-mayeur) [inline src/models/login.model.js]: I don't understant why you write `="secret"` ?

Comment (TheoPth) [reply src/models/login.model.js]: It's the secret to encode the jwt token. It can be anything.

Comment (TheoPth) [reply src/models/login.model.js]: Oh no, it's because i made wrong

Comment (yannick-mayeur) [reply src/models/login.model.js]: fix une boulette :laughing:

**Comment by yannick-mayeur** [comment]
Pour moi c'est mergeable, mais il faut qu'on voit comment pouvoir check l'auth sur les ressources protégés par la connexion.

**Pull request merged**